File: app/src/main/java/com/icerockdev/babenko/core/rx/RxUtils.java
package com.icerockdev.babenko.core.rx;

import io.reactivex.ObservableTransformer;
import io.reactivex.SingleTransformer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by Roman Babenko on 17/10/17.
 */

public class RxUtils {
    public static <T> ObservableTransformer<T, T> applyIoMainThreadSchedulersToObservable() {
        return upstream -> upstream
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public static <T> SingleTransformer<T, T> applyIoMainThreadSchedulersToSingle() {
        return upstream -> upstream
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }
}

File: app/src/main/java/com/icerockdev/babenko/utils/RxUtils.kt
package com.icerockdev.babenko.utils

import io.reactivex.ObservableTransformer
import io.reactivex.SingleTransformer
import io.reactivex.schedulers.Schedulers
import io.reactivex.android.schedulers.AndroidSchedulers;


/**
 * Created by Roman Babenko on 17/10/17.
 */

object RxUtils {
    fun <T> applyIoMainThreadSchedulersToObservable(): ObservableTransformer<T, T> {
        return ObservableTransformer { upstream ->
            upstream
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
        }
    }

    fun <T> applyIoMainThreadSchedulersToSingle(): SingleTransformer<T, T> {
        return SingleTransformer { upstream ->
            upstream
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
        }
    }
}
